torch/csrc/distributed/rpc/process_group_agent.cpp

```diff
@@ -297,6 +294,29 @@ std::shared_ptr<FutureMessage> ProcessGroupAgent::send(
     future->markCompleted();
   }
 
+  // Sending to ourselves: bypass the send logic and enqueue directly
+  // to our receving queue.
+  if (to.id_ == (worker_id_t)pg_->getRank()) {
+    TORCH_CHECK(!message.isShutdown(), "Shutting down self not supported");
+    threadPool_.run(std::bind(
+        [this](const Message& message) {
+          sendCounts_.increment(pg_->getRank());
+          // Unlike the other cases, need to add a tensor deleter, since the
+          // data outlives the scope of this function.
+          auto serializedPayload = new std::string(serialize(message));
+          enqueueRecv(RecvWork(
+              allWorkerInfo_[pg_->getRank()],
+              message.type(),
+              torch::from_blob(
+                  (void*)serializedPayload->data(),
+                  serializedPayload->length(),
+                  [serializedPayload](void*) { delete serializedPayload; },
```

Comment:
This is capture by value? Does this prevent the local var `serializedPayload` being freed before the recv is done?

Reply:
serializedPayload is a string*, so we capture the pointer by value.
The goal is for the deleter can delete the buffer when the torch::Tensor is done with it.
The local var won't clean itself up.

In the other torch::from_blob() cases in this file, the tensor buffer doesn't need to be persisted beyond the immediate function scope. This case is different because the Tensor's use is delayed until the other listen thread picks it up.

Maybe I'll go ahead and make it a unique_ptr<>, to make the code a bit more exception-safe, and to clarify the lifetime.

Reply:
I went ahead and uploaded a version with a smart pointer to make memory lifetime clearer in this func, and safer with exceptions.

That said, I had to use shared_ptr<> instead of unique_ptr<> because Pytorch is supposed to avoid c++14 code (TIL there are some c++11 issues with unique_ptr<> lambda capture that were solved in c++14)

Reply:
@jjlilley Thanks to @smessmer, we can use  C++14 since 2 weeks (see #28443).

Capturing a `std::unique_ptr` is possible only if you don't convert it to `std::function`, because that requires the wrapped function to be copy-constructible. And of course, if you capture a `std::unique_ptr`, it won't be copy-constructible.

Reply:
Do we then want to wait for #27864?

```
Nov 07 01:02:43 /var/lib/jenkins/workspace/torch/csrc/distributed/rpc/process_group_agent.cpp: In lambda function:
Nov 07 01:02:43 /var/lib/jenkins/workspace/torch/csrc/distributed/rpc/process_group_agent.cpp:316:20: error: lambda capture initializers only available with -std=c++14 or -std=gnu++14 [-Werror]
Nov 07 01:02:43                    [payload = std::move(payload)](void*) {
```

Reply:
actually, if I create a unique_ptr<> and release the raw pointer into the deleter lambda captures, I think we both:
  1. are still fairly exception-safe, and mostly avoid raw pointers
  2. self-document the memory lifecycle reasonably
I'll upload a version that does this.